llvm/tools/sycl-post-link/ModuleSplitter.h

```diff
@@ -126,18 +126,43 @@ class ModuleDesc {
              const Properties &Props)
       : M(std::move(M)), EntryPoints(std::move(EntryPoints)), Props(Props) {
     Name = this->EntryPoints.GroupId.str();
+    fixupEntryPointsLinkage();
   }
 
   ModuleDesc(std::unique_ptr<Module> &&M, const std::vector<std::string> &Names,
              StringRef Name = "NoName")
       : M(std::move(M)), Name(Name) {
     rebuildEntryPoints(Names);
+    fixupEntryPointsLinkage();
+  }
+
+  // Some entry points may have linkonce_odr linkage, which does not really work
+  // for an entry point, because either inliner pass or LLVM linker will remove
+  // functions with such linkage, and entry point must be always available by
+  // definition. This function changes linkage of linkonce_odr entry points.
+  // An example of a such entry point is an instantiation of a function template
+  // even marked SYCL_EXTERNAL.
+  // TODO: is it practical to fix FE instead to assign different linkage for
+  // SYCL_EXTERNAL functions which get linkonce_odr currently?
+  bool fixupEntryPointsLinkage() {
+    bool Changed = false;
+
+    for (Function *F : EntryPoints.Functions) {
+      GlobalValue::LinkageTypes L = F->getLinkage();
+
+      if ((L == GlobalValue::LinkageTypes::LinkOnceODRLinkage) ||
+          (L == GlobalValue::LinkageTypes::LinkOnceAnyLinkage)) {
+        F->setLinkage(GlobalValue::LinkageTypes::ExternalLinkage);
```

Comment:
Why do you change it to external and not weak_odr?

Reply:
Entry point need to be available externally, yet an attempt to later link in another module which defines the same entry point would be an error. So `ExternalLinkage` seems the best fit.

Reply:
Why would it be an error to have the same template function (which are entry point to invoke_simd) in two different TUs? This is valid C++. Why should this not work for invoke_simd?